src/core/internal/gc/impl/conservative/gc.d

```diff
@@ -2464,19 +2470,25 @@ struct Gcx
                                     void* q = sentinel_add(p);
                                     sentinel_Invariant(q);
 
-                                    if (pool.finals.nbits && pool.finals.test(biti))
-                                        rt_finalizeFromGC(q, sentinel_size(q, size), pool.getBits(biti));
+                                    static if (hasFinalizer)
+                                        if (pool.finals.nbits && pool.finals.test(biti))
+                                            rt_finalizeFromGC(q, sentinel_size(q, size), pool.getBits(biti));
 
                                     assert(core.bitop.bt(toFree.ptr, i));
 
                                     debug(COLLECT_PRINTF) printf("\tcollecting %p\n", p);
-                                    leakDetector.log_free(sentinel_add(p));
+                                    leakDetector.log_free(sentinel_add(p), sentinel_size(q, size));
 
                                     debug (MEMSTOMP) memset(p, 0xF3, size);
                                 }
                             }
                         }
 
+                        if (hasFinalizer)
+                            sweepBin!true();
+                        else
+                            sweepBin!false();
```

Comment:
Does this compile to nothing if all debug options are off? I doubt that, so maybe prepend `debug`.

Reply:
In DMD, it doesn't compile to nothing. I think it should compile to nothing in LDC/GDC but I haven't checked yet.

`debug` isn't quite right, there's a few conditions affecting this code.

Reply:
> In DMD, it doesn't quite compile to nothing.

I think that would be a blocker for adding the else path unconditionally. An appropriate compile time condition should be possible.

Reply:
I suspect it would be inobservable in benchmarks, because the loop just increments registers.

But, I think it would be better to just split this into several loops, each with a top-level conditional.

Reply:
> I think it should compile to nothing in LDC/GDC, but I haven't checked.

It does not:

https://forum.dlang.org/post/aejdlmkvuwgbtxktpjby@forum.dlang.org

Reply:
> But, I think it would be better to just split this into several loops, each with a top-level conditional.

Done.

The best solution I could come up with here is a string mixin.